File: src/structures/Member.ts
import { Base, Role } from '.'
import { validator } from '../utils'
import sql from '../database'
import config from '../config'

export interface CreateMemberOptions extends Options<Member> {
  id: string
  server_id: string
}

export const UpdateMemberSchema = validator.compile({
  nickname: `string|min:0|max:${config.limits.member.nickname}|nullable`,
  roles: 'snowflake[]|unique|optional'
})


export class Member extends Base {
  nickname: Nullable<string> = null
  joined_at = Date.now()
  server_id!: string
  roles: string[] = []

  fetchRoles(): Promise<Role[]> {
    return Role.find({ id: this.roles })
  }

  static from(opts: CreateMemberOptions): Member {
    return Object.assign(new Member(), opts)
  }
}

File: src/structures/member.rs
use super::*;
use crate::utils::Snowflake;
use chrono::{NaiveDateTime, Utc};
use ormlite::model::*;
use serde::{Deserialize, Serialize};

#[derive(Debug, OpgModel)]
struct MemberRoles(Vec<String>);

#[serde_as]
#[derive(Debug, Serialize, Deserialize, Model, FromRow, Clone, OpgModel)]
#[ormlite(table = "members")]
pub struct Member {
    pub id: Snowflake,
    pub nickname: Option<String>,
    #[opg(string)]
    pub joined_at: NaiveDateTime,
    pub server_id: Snowflake,
    #[opg(custom = "MemberRoles")]
    pub roles: Vec<Snowflake>,
}

impl Member {
    pub fn new(user_id: Snowflake, server_id: Snowflake) -> Self {
        Self {
            id: user_id,
            nickname: None,
            server_id,
            roles: vec![server_id],
            joined_at: Utc::now().naive_utc(),
        }
    }

    pub async fn fetch_roles(&self) -> Result<Vec<Role>, ormlite::Error> {
        if self.roles.is_empty() {
            return Ok(vec![]);
        }

        Role::select()
            .filter("id = ANY($1) AND server_id = $2")
            .bind(self.roles.clone())
            .bind(self.server_id)
            .fetch_all(pool())
            .await
    }

    #[cfg(test)]
    pub async fn faker() -> Result<Self, Error> {
        let server = Server::faker().await?;
        let member = Self::new(server.owner_id, server.id);

        server.save().await?;

        Ok(member)
    }
}

impl Base for Member {}
